aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java

```diff
@@ -0,0 +1,150 @@
+import java.io.IOException;
+import java.util.Arrays;
+import org.apache.iceberg.io.SeekableInputStream;
+import org.apache.iceberg.relocated.com.google.common.base.Joiner;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class S3InputStream extends SeekableInputStream {
+  private static final Logger LOG = LoggerFactory.getLogger(S3InputStream.class);
+
+  private final StackTraceElement[] createStack;
+  private final AmazonS3 s3;
+  private final AmazonS3URI location;
+
+  private S3ObjectInputStream stream;
+  private long pos = 0;
+  private long next = 0;
+  private boolean closed = false;
+
+  private int skipSize = 1024 * 1024;
+
+  public S3InputStream(AmazonS3 s3, AmazonS3URI location) {
+    this.s3 = s3;
+    this.location = location;
+
+    createStack = Thread.currentThread().getStackTrace();
+  }
+
+  @Override
+  public long getPos() {
+    return pos;
+  }
+
+  @Override
+  public void seek(long newPos) {
+    Preconditions.checkState(!closed, "already closed");
+    Preconditions.checkArgument(newPos >= 0, "position is negative: %s", newPos);
+
+    // this allows a seek beyond the end of the stream but the next read will fail
+    next = newPos;
+  }
+
+  @Override
+  public int read() {
+    // This stream is wrapped with BufferedInputStream, so this method should never be called
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public int read(byte[] b, int off, int len) throws IOException {
+    Preconditions.checkState(!closed, "already closed");
+    positionStream();
+
+    int bytesRead = stream.read(b, off, len);
+    pos += bytesRead;
+    next += bytesRead;
+
+    return bytesRead;
+  }
+
+  @Override
+  public void close() throws IOException {
+    super.close();
+    closed = true;
+  }
+
+  private void positionStream() throws IOException {
+    if ((stream != null) && (next == pos)) {
+      // already at specified position
+      return;
+    }
+
+    if ((stream != null) && (next > pos)) {
+      // seeking forwards
+      long skip = next - pos;
+      if (skip <= Math.max(stream.available(), skipSize)) {
+        // already buffered or seek is small enough
+        LOG.debug("Read-through seek for {} to offset {}", location, next);
+        try {
+          if (stream.skip(skip) == skip) {
+            pos = next;
+            return;
+          }
+        } catch (IOException ignored) {
+          // will retry by re-opening the stream
+        }
+      }
+    }
+
+    // close the stream and open at desired position
+    LOG.debug("Seek with new stream for {} to offset {}", location, next);
+    pos = next;
+    closeStream();
+    openStream();
+  }
+
+  private void openStream() {
+    GetObjectRequest request = new GetObjectRequest(location.getBucket(), location.getKey())
+        .withRange(pos, Long.MAX_VALUE);
+    stream = s3.getObject(request).getObjectContent();
```

Comment:
Should a `null` check be added here to ensure that the previous stream was closed / aborted? I see that you're using this function immediately after calling `closeStream`, which has the null check. However, I worry that in a future refactor that tightly coupling of thnull check between `closeStream` and `openStream` will be lost and then resources will be potentially leaked.

Reply:
So, I changed this a little bit so that the closing is handled just prior to opening the new stream, so that there's less chance it would be leaked in the future.  I also updated to not call the `sream.abort()` because that is actually handled by the underlying stream and will abort if necessary.  Now we are just calling `close()`, which is probably more standard.